### PR TITLE
#594 Implement check to prohibit Javadoc in methods with '@Override'

### PR DESCRIPTION
**Pull request by HDouss**
Resolving #594.
Added implementation and unit tests.

### CODE REVIEW

**Comment by davvd** [comment]
@hdouss Many thanks, I will find someone to review it before we merge

**Comment by davvd** [comment]
@mkordas it's yours,please review

**Comment by mkordas** [comment]
@HDouss I'll review it later today

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/NoJavadocForOverrideMethods.java [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+import com.puppycrawl.tools.checkstyle.utils.AnnotationUtility;
+
+/**
+ * Checks that there is no javadoc for inherited methods.
+ *
+ * @author Hamdi Douss (douss.hamdi@gmail.com)
+ * @version $Id$
+ * @since 0.15.5
+ */
+public final class NoJavadocForOverrideMethods extends Check {
```
Comment (mkordas) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/NoJavadocForOverrideMethods.java]: @HDouss please follow convention and add to this name `Check` at the end

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/NoJavadocForOverrideMethods.java [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+    @Override
+    public void visitToken(final DetailAST ast) {
+        if (AnnotationUtility.containsAnnotation(ast, "Override")) {
+            final FileContents contents = getFileContents();
+            final TextBlock javadoc = contents.getJavadocBefore(
+                ast.getLineNo()
+            );
+            if (javadoc != null) {
+                log(ast, "Overriden methods should not have javadoc");
```
Comment (mkordas) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/NoJavadocForOverrideMethods.java]: @HDouss typo: should be `Overridden`

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/NoJavadocForOverrideMethods.java (hunk 1 above)
Comment (mkordas) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/NoJavadocForOverrideMethods.java]: @HDouss you need to be consistent in naming and decide either to "override methods" or "overridden methods"

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverrideMethods/Invalid.java [hunk 3]
```diff
@@ -0,0 +1,29 @@
+/**
+ * This is not a real Java class. It won't be compiled ever. It is used
+ * only as a text resource in integration.ChecksIT.
+ */
+
+public final class Valid {
+
+    @Override
+    public void overriden() {
```
Comment (mkordas) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverrideMethods/Invalid.java]: @HDouss typo in method name

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverrideMethods/Invalid.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
+/**
+ * This is not a real Java class. It won't be compiled ever. It is used
+ * only as a text resource in integration.ChecksIT.
+ */
+
+public final class Valid {
```
Comment (mkordas) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverrideMethods/Invalid.java]: @HDouss class should be rather named `Invalid`

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverrideMethods/Invalid.java [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void inherited() {
+    }
+
+    @Override
+    public void good() {
```
Comment (mkordas) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverrideMethods/Invalid.java]: @HDouss I don't think we need good example in "invalid" class

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverrideMethods/Valid.java [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+    public void main() {
+    }
+
+    /**
+     * Javadoc.
+     */
+    public void note() {
+    }
+
+    @Override
```
Comment (mkordas) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverrideMethods/Valid.java]: @HDouss can you also verify that normal comments like `//` or `/*` are still allowed above such methods?

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/NoJavadocForOverrideMethods.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+import com.puppycrawl.tools.checkstyle.api.Check;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.FileContents;
+import com.puppycrawl.tools.checkstyle.api.TextBlock;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.AnnotationUtility;
+
+/**
+ * Checks that there is no javadoc for inherited methods.
```
Comment (mkordas) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/NoJavadocForOverrideMethods.java]: @HDouss Here we should provide a reason why it is a bad practice.

**Comment by HDouss** [comment]
@mkordas Thanks. Pushed.

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java [hunk 8]
```diff
@@ -0,0 +1,26 @@
+/**
+ * This is not a real Java class. It won't be compiled ever. It is used
+ * only as a text resource in integration.ChecksIT.
+ */
+
+public final class Invalid {
+
+    @Override
+    public void overridden() {
```
Comment (mkordas) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java]: @HDouss this one is valid too, can be removed

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/NoJavadocForOverriddenMethodsCheck.java [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+
+    @Override
+    public void visitToken(final DetailAST ast) {
+        if (AnnotationUtility.containsAnnotation(ast, "Override")) {
+            final FileContents contents = getFileContents();
+            final TextBlock javadoc = contents.getJavadocBefore(
+                ast.getLineNo()
+            );
+            if (javadoc != null) {
+                log(ast, "Overridden methods should not have javadoc");
```
Comment (mkordas) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/NoJavadocForOverriddenMethodsCheck.java]: @HDouss should be `Javadoc` - capital letter

File: qulice-checkstyle/src/main/java/com/qulice/checkstyle/NoJavadocForOverriddenMethodsCheck.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+
+import com.puppycrawl.tools.checkstyle.api.Check;
+import com.puppycrawl.tools.checkstyle.api.DetailAST;
+import com.puppycrawl.tools.checkstyle.api.FileContents;
+import com.puppycrawl.tools.checkstyle.api.TextBlock;
+import com.puppycrawl.tools.checkstyle.api.TokenTypes;
+import com.puppycrawl.tools.checkstyle.utils.AnnotationUtility;
+
+/**
+ * Checks that there is no javadoc for inherited methods.
```
Comment (mkordas) [inline qulice-checkstyle/src/main/java/com/qulice/checkstyle/NoJavadocForOverriddenMethodsCheck.java]: @HDouss should be `Javadoc` - capital letter

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java [hunk 11]
```diff
@@ -0,0 +1,26 @@
...
+    public void javadoc() {
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void inherited() {
+    }
+
```
Comment (mkordas) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java]: @HDouss can you also make a negative test with one-line Javadoc?

**Comment by mkordas** [comment]
@HDouss just a couple of final comments and we are done

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java (hunk 11 above)
Comment (HDouss) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java]: @mkordas How is a one line Javadoc looks different from the first method Javadoc ?

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java (hunk 8 above)
Comment (HDouss) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java]: @mkordas I don't think it is problematic to have a valid case here. But I'll do it like the other one.

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java (hunk 11 above)
Comment (HDouss) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java]: @mkordas I don't know if it is what you mean, but I made an additional method with: `/** Javadoc. */`

**Comment by HDouss** [comment]
@mkordas Done.

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java (hunk 11 above)
Comment (mkordas) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java]: @HDouss yes, that's exactly what I meant

File: qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java (hunk 8 above)
Comment (mkordas) [inline qulice-checkstyle/src/test/resources/com/qulice/checkstyle/ChecksTest/NoJavadocForOverriddenMethodsCheck/Invalid.java]: @HDouss not problematic, but for me it was a bit confusing

**Comment by mkordas** [comment]
@HDouss looks good!

**Comment by mkordas** [comment]
@rultor merge

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by HDouss** [comment]
@krzyk Is there some problem or should we just trigger another merge ?

**Comment by krzyk** [comment]
@HDouss it means that you have to merge master changes

**Comment by HDouss** [comment]
@krzyk Strange. I have merged. I will retry.

**Comment by mkordas** [comment]
@krzyk why do you think master should be merged? GitHub show the branch as mergeable

**Comment by HDouss** [comment]
@krzyk I confirm that the branch is up to date.

**Comment by HDouss** [comment]
@krzyk Can you try to merge again ?

**Comment by krzyk** [comment]
@HDouss there are cases when github says it is OK, but rultor can't merge, it usually informs with such messages as above, but let's try

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by krzyk** [comment]
@HDouss strange, it still insists on merging master changes

**Comment by HDouss** [comment]
@krzyk Already up to date :/

**Comment by HDouss** [comment]
@krzyk The weird thing is that `ChecksTest/NoJavadocForOverriddenMethodsCheck/violations.txt` does not exist in master.

**Comment by krzyk** [comment]
@yegor256 could you help us out here? rultor has problems merging this branch

**Comment by mkordas** [comment]
@HDouss I know the reason. You've committed `violations.txt` with CRLF instead of LF and Rultor doesn't handle that unfortunately.

**Comment by mkordas** [comment]
@HDouss have you tried committing this file again with just LF?

**Comment by HDouss** [comment]
@mkordas Thanks !

**Comment by HDouss** [comment]
@krzyk Please reask for merge.

**Comment by krzyk** [comment]
@rultor merge pls

**Comment by mkordas** [comment]
@rultor deploy

**Comment by davvd** [comment]
@mkordas Many thanks! **37 mins** were added to your account in Transaction ID `56a7591940ca84178a0002a8` (task took 72 hours and 7 mins)... extra minutes for review comments (c=22)... added +37 to your rating, now it is equal to [+3579](http://www.netbout.com/b/36080?open=rating)

**Comment by davvd** [comment]
@rultor deploy now pls

**Pull request merged**